astropy/io/fits/tests/test_image.py

```diff
@@ -73,7 +73,7 @@ def test_constructor_copies_header(self):
         # Original header should be unchanged
         assert phdr['FILENAME'] == 'labq01i3q_rawtag.fits'
 
-    @raises(ValueError)
+    @raises(IndexError)
```

Comment:
Yes I agree it could be done, I'm just not sure of the best way. And `self._read_all` should already give this information, so is it better to add a new way to know if the file was fully read ? Instead why not simply removing [the line](https://github.com/astropy/astropy/blob/master/astropy/io/fits/hdu/hdulist.py#L1146) that set `self._read_all` to True when the file is closed ? I guess this line is there to avoid running again the beginnning of `_read_next_hdu`, but if the file is closed it's possible anyway to return earlier.